data/blog/id-headers-tailscale-serve-flask.mdx

```diff
@@ -0,0 +1,79 @@
+---
+title: Tapping into Tailscale's identity headers with serve
+date: 2023-07-19
+authors: ['parker', 'sonia']
+tags:
+  - serve
+  - headers
+  - python
+  - flask
+  - funnel
+summary: When you use Tailscale serve to proxy requests, you can use its identity headers to change what content you serve.
+images: ['/images/id-headers-tailscale-serve-flask/social.png']
+---
+
+If you’ve got multiple users on your tailnet, or if you’ve shared a node out to multiple tailnets, you may want to serve different content based on who’s asking. For personal uses, though, implementing an actual log-in system is cumbersome and tricky to persist across devices.
```

Comment:
I have a few concerns about the opening paragraph. It's important that we quickly touch on Funnel and Serve and link to previous blog posts. (There's also a related tsnet post on .com https://tailscale.com/blog/tsnet-virtual-private-services/)

Something like "You may have heard of [Funnel], a secure tunnel which allows you to share local services over the internet. But did you know you can also share privately? Tailscale Serve is a private sharing feature, and in fact, it's what Funnel connects to. The two features, Funnel and Serve, work together to serve public traffic from the Internet and over your tailnet."

Reply:
```suggestion
You may have heard about [Funnel](https://tailscale.com/kb/1247/funnel-serve-use-cases/), a secure tunnel that lets you share local services over the internet; but did you know you can also share those services privately? Tailscale Serve lets you expose local services to your tailnet (and in fact is what Funnel connects to). Funnel and Serve work together to serve public traffic from the internet and private traffic from your tailnet. However, once you have everything wired up, your application is going to need to figure out who is on the other end of the connection. Implementing login systems like this are fraught with peril and easily one of the most complicated things that us programmers have to implement over and over and over again.

However, let's take a moment to step back and think about the parts we are using here. We're connected over Tailscale. In order to authenticate with Tailscale, you have to pass a two-factor auth check, and your identity provider tells Tailscale who you are. This means there's a mapping of IP addresses to names so you can _completely bypass_ the need to find new and interesting ways to tell your application who the client is because Tailscale already knows that. Today we're going to be talking about how you can use Tailscale as your authentication system for internal services with the magic of HTTP headers.
```